pyaerocom/colocation_3d.py

```diff
@@ -0,0 +1,511 @@
+"""
+Methods and / or classes to perform 3D colocation
+"""
+from __future__ import annotations
+
+import logging
+import os
+from typing import NamedTuple
+
+import iris
+import numpy as np
+import pandas as pd
+import xarray as xr
+from cf_units import Unit
+
+from pyaerocom import __version__ as pya_ver
+from pyaerocom import const
+from pyaerocom.colocateddata import ColocatedData
+from pyaerocom.colocation import (
+    _colocate_site_data_helper,
+    _colocate_site_data_helper_timecol,
+    _regrid_gridded,
+    check_time_ival,
+    check_ts_type,
+    resolve_var_name,
+)
+from pyaerocom.exceptions import (
+    DataUnitError,
+    DimensionOrderError,
+    MetaDataError,
+    TemporalResolutionError,
+    TimeMatchError,
+    VariableDefinitionError,
+    VarNotAvailableError,
+)
+from pyaerocom.filter import Filter
+from pyaerocom.helpers import make_datetime_index, to_pandas_timestamp
+from pyaerocom.time_resampler import TimeResampler
+from pyaerocom.tstype import TsType
+from pyaerocom.variable import Variable
+
+logger = logging.getLogger(__name__)
+
+
+class ColocatedDataLists(NamedTuple):
+    colocateddata_for_statistics: list[ColocatedData]
+    colocateddata_for_profile_viz: list[ColocatedData]
+
+
+def _colocate_vertical_profile_gridded(
+    data,
+    data_ref,
+    ts_type=None,
+    start=None,
+    stop=None,
+    filter_name=None,
+    regrid_res_deg=None,
+    harmonise_units=True,
+    regrid_scheme="areaweighted",
+    var_ref=None,
+    update_baseyear_gridded=None,
+    min_num_obs=None,
+    colocate_time=False,
+    use_climatology_ref=False,
+    resample_how=None,
+    layer_limits=None,
+    **kwargs,
+) -> list[ColocatedData]:
+    if layer_limits is None:
+        raise Exception(f"layer limits must be provided")
+
+    obs_stat_data = kwargs["obs_stat_data"]
```

Comment:
why are these kwargs and not normal keywords?

Reply:
Good point. Changed

Reply:
Changed